### PR TITLE
Clarifying Authentication Features

### PR DESCRIPTION
**Pull request by robertjd**
This PR clarifies the following areas:
- How we use our OAuth2 access & refresh tokens for browser authentication
- How our Oauth2 features should be exposed as spec-compliant features through the web application's `/oauth/token` endpoint
- The purpose of the `/me` route

I'm closing #6 in favor of this one, as it addresses all the issues that were brought up in that PR

### CODE REVIEW

**Comment by robertjd** [comment]
Ping @lhazlewood for review.  Of important is the `validationStrategies` option in the `oauth2` section.  Thanks!

File: cookie-authentication.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  * Othwerwise, use `web.domain` if defined
+  * Finally, fallback to "" as the value
+
+* The `secure` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.secure` is defined, use that value
+  * Othwerwise, the framework should make a best effort to determine if the
+    incoming request is HTTPS and set the `secure` flag if so.
+
+## <a name="Configuration Options"></a> Configuration Options
```
Comment (timothyej) [inline cookie-authentication.md:81]: There's no need for the html `a` tag - GitHub will automatically create it for you based on the title, in this case `#configuration-options` :)

**Comment by lhazlewood** [comment]
@robertjd per our discussion this morning, you indicated you might split this up to represent what users can config vs. what our internal implementations should do for certain values.  Do you still want me to review or should I wait?

**Comment by robertjd** [comment]
I do want to clarify those points, but can you please take a look at the `validationStrategies` ?

**Comment by robertjd** [comment]
@lhazlewood I've updated this PR per our coversation the other day.

File: cookie-authentication.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+* The `HttpOnly` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.httpOnly` is false, do not set this flag
+  * Othwerwise, set this flag to true
+
+* The `path` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.path` is defined, use that value
+  * Othwerwise, use `web.basePath` if defined
+  * Finally, fallback to "/" as the value
```
Comment (lhazlewood) [inline cookie-authentication.md]: I think this should be read as:

> If `stormpath.web.[accessTokenCookie|refreshTokenCookie].path` has a non-null value, use that value as the cookie path.  If the value is undefined or null, default to the web application framework default value.  If the framework does not provide a default value, fall back to `stormpath.web.basePath` if defined.  If `stormpath.web.basePath` is undefined or null, fallback to `/`.

Backstory:

Every Java web application has what is called the [context path](http://docs.oracle.com/javaee/7/api/javax/servlet/ServletContext.html#getContextPath--) - the base path off of root where your application lives.  If you set a cookie and do not specify a cookie path, the Java container automatically sets the cookie path to be your web app's `contextPath`.

The `contextPath` satisfies the behavior proposed by having a `stormpath.web.basePath` property.  As such, this implies there is never a need to configure a `stormpath.web.basePath` in a Java web app because that is always available.

The Java `contextPath` can differ depending on how you deploy your app to the java web/application server.  You can deploy it to the root of a domain, such that `/` becomes the context path, or you can deploy it to `/foo` and that becomes the context path.

In other words, what base path your app is deployed to is an operations/deployment-time concern and not something that a Java app developer should be required to configure.  The Java developer always references `request.getServletContext().getContextPath()` when coding so their app always works, regardless of what path structure exists at runtime.

This allows Java devs to deploy their app to `http://localhost:8080/` in development and have it work in production under `https://company.com/foo/bar` without changing any code or configuration.

File: cookie-authentication.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+* The `domain` flag should be set as follows:
+
+  * If `web.<accessTokenCookie|refreshTokenCookie>.domain` is defined, use that value
```
Comment (lhazlewood) [inline cookie-authentication.md]: A `null` value would indicate 'let the web framework set the domain`.

I'm not sure if `web.domain` is required.

File: cookie-authentication.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+in these option blocks.  But you will need to know what those policy settings
+are, and you will find them at `web.oauth2.password`.
+
+```yaml
+web:
+  accessTokenCookie:
+    name: "access_token"
+    httpOnly: true
+    secure: null
+    path: "/"
```
Comment (lhazlewood) [inline cookie-authentication.md]: `path: null` (per comments above about `contextPath`).  Our heuristics in a node app can fall back to `/` if nothing else could be found - no need to require a concrete value in the config file.

File: cookie-authentication.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    domain: null
+  refreshTokenCookie:
+    name: "refresh_token"
+    httpOnly: true
+    secure: null
+    path: "/"
```
Comment (lhazlewood) [inline cookie-authentication.md]: `path: null`

File: oauth2.md [hunk 6]
```diff
@@ -2,65 +2,164 @@
...
+  "token_type":"Bearer"
+}
+```
+
+### Client Credentials Options
+
+```yaml
+web:
+  oauth2:
     client_credentials:
```
Comment (lhazlewood) [inline oauth2.md:82]: Thinking out loud here, but not entirely sure...

Instead of `client_credentials:  enabled: true`, what about:

``` yaml
web:
    oauth2:
        grantTypes:
            - client_credentials:
                  enabled: true
            - password
```

i.e. a list?  or should it still be name-value pairs?  Thoughts?

Also, should each grant type have its own ttl setting?  Or would that be shared:

``` yaml
web:
    oauth2:
        ttl: 3600
        grantTypes:
            - client_credentials:
                  enabled: true
            - password
```

?

File: web-config.yaml [hunk 7]
```diff
@@ -1,5 +1,7 @@
 ---
 web:
+  domain: null
+  basePath: null
```
Comment (lhazlewood) [inline web-config.yaml:3]: In a Java web app, `basePath` should never be set.  Should this (and `domain`) be a required configuration field?

File: web-config.yaml [hunk 8]
```diff
@@ -9,17 +11,18 @@ web:
         ttl: 3600
     password:
       enabled: true
+      validationStrategies: ['local','stormpath']
   accessTokenCookie:
     name: "access_token"
     httpOnly: true
     secure: null
-    path: "/"
+    path: null
```
Comment (lhazlewood) [inline web-config.yaml:23]: Interesting - you already made the path change here.  I suspect this is the current incarnation `path: null` and the above was just out of date?

File: web-config.yaml (hunk 7 above)
Comment (robertjd) [inline web-config.yaml:3]: Per other comments, I've removed `domain` but left `basePath`, as some frameworks don't have a canonical way to declare a "context path" (Express is one).

File: oauth2.md (hunk 6 above)
Comment (robertjd) [inline oauth2.md:82]: We need to have separate ttl settings, until we add the `client_credentials` workflow to the core API and both workflows share the OAuth policy that is defined by the application.

And I think these should remain as property maps with `enabled`, as it's consistent with the rest of the spec and I'm finding this structure very flexible for layering configuration options, per our configuration loading heuristics.
